### PR TITLE
Allow for relative attention models to clip the relative positions during evaluation

### PR DESCRIPTION
**Pull request by jxihong**

### CODE REVIEW

File: latent_programmer/decomposition_transformer_attention/decomposition_models.py [hunk 1]
```diff
@@ -90,11 +92,45 @@ class DecomposeAttentionTransformer(nn.Module):
...
+      base_encoder_config = base_config
+      base_decoder_config = base_config.replace(shift=False)
+
+    self.encoder = base_models.TransformerIOEncoder(
+        config=base_encoder_config, name='encoder')
     self.decoder = base_models.TransformerDecoder(
-        config=base_config.replace(shift=False), name='decoder')
+        config=base_decoder_config, name='decoder')
+
+    self.max_input_distance = self.variable(
```
Comment (kensens) [inline latent_programmer/decomposition_transformer_attention/decomposition_models.py:126]: can we put this inside an `if not is_initialized:` block? we don't want to overwrite them if they're already initialized

File: latent_programmer/decomposition_transformer_attention/decomposition_models.py [hunk 2]
```diff
@@ -223,4 +258,22 @@ def __call__(self,
     encoded = self.encode(inputs, outputs)
     encoded_padding_mask = jnp.where(outputs > 0, 1, 0).astype(jnp.float32)
 
+    if self.has_variable('relative_attention', 'max_input_distance'):
+      max_input_distance = jnp.max(jnp.count_nonzero(inputs == 0, axis=0))
```
Comment (kensens) [inline latent_programmer/decomposition_transformer_attention/decomposition_models.py:262]: is `jnp.count_nonzero(array == 0, axis=0)` correct? i think instead we want `jnp.count_nonzero(array != 0, axis=1) - 1`, where the -1 is because we want distance not the number of tokens. but for programs, we need -2 since the last element is EOS which won't produce a distance

File: latent_programmer/decomposition_transformer_attention/train.py [hunk 3]
```diff
@@ -241,43 +259,55 @@ def train_step(optimizer,
...
   new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
 
   # Get metrics.
   metrics = compute_metrics(logits, programs, weights)
   metrics['learning_rate'] = lr
-  return new_optimizer, metrics, new_dropout_rng
+  new_state = state.replace(
+    step=step + 1,
+    optimizer=new_optimizer,
+    ra_stats=jax.lax.pmean(ra_stats, 'batch'))
```
Comment (kensens) [inline latent_programmer/decomposition_transformer_attention/train.py:287]: do we want pmean here? don't we need a pmax instead?

**Review by kensens — COMMENTED** [review]
I left 3 comments suggesting changes. If I understand the code correctly, then you don't need to make the changes here in the PR, because I'll make them directly in my CL (which is easier for both of us since I already have the CL)
